libs/partners/azure-dynamic-sessions/langchain_azure_dynamic_sessions/tools/sessions.py

```diff
@@ -0,0 +1,256 @@
+import importlib.metadata
+import os
+import re
+import urllib
+from dataclasses import dataclass
+from datetime import datetime, timedelta, timezone
+from io import BufferedReader, BytesIO
+from typing import Any, Callable, Optional
+from uuid import uuid4
+
+import requests
+from azure.core.credentials import AccessToken
+from azure.identity import DefaultAzureCredential
+from langchain_core.runnables.config import run_in_executor
+from langchain_core.tools import BaseTool
+
+try:
+    _package_version = importlib.metadata.version("langchain-azure-dynamic-sessions")
+except importlib.metadata.PackageNotFoundError:
+    _package_version = "0.0.0"
+USER_AGENT = f"langchain-azure-dynamic-sessions/{_package_version} (Language=Python)"
+
+
+def _access_token_provider_factory() -> Callable[[], Optional[str]]:
+    """Factory function for creating an access token provider function.
+
+    Returns:
+        Callable[[], Optional[str]]: The access token provider function
+    """
+
+    access_token: AccessToken = None
+
+    def access_token_provider() -> Optional[str]:
+        nonlocal access_token
+        if access_token is None or datetime.fromtimestamp(
+            access_token.expires_on, timezone.utc
+        ) < datetime.now(timezone.utc) + timedelta(minutes=5):
+            credential = DefaultAzureCredential()
+            access_token = credential.get_token("https://dynamicsessions.io/.default")
+        return access_token.token
+
+    return access_token_provider
+
+
+def _sanitize_input(query: str) -> str:
+    """Sanitize input to the python REPL.
+    Remove whitespace, backtick & python (if llm mistakes python console as terminal)
+
+    Args:
+        query: The query to sanitize
+
+    Returns:
+        str: The sanitized query
+    """
+
+    # Removes `, whitespace & python from start
+    query = re.sub(r"^(\s|`)*(?i:python)?\s*", "", query)
+    # Removes whitespace & ` from end
+    query = re.sub(r"(\s|`)*$", "", query)
+    return query
+
+
+@dataclass
+class RemoteFileMetadata:
+    """Metadata for a file in the session."""
+
+    filename: str
+    """The filename relative to `/mnt/data`."""
+
+    size_in_bytes: int
+    """The size of the file in bytes."""
+
+    @property
+    def full_path(self) -> str:
+        """Get the full path of the file."""
+        return f"/mnt/data/{self.filename}"
+
+    @staticmethod
+    def from_dict(data: dict) -> "RemoteFileMetadata":
+        """Create a RemoteFileMetadata object from a dictionary."""
+        properties = data.get("properties", {})
+        return RemoteFileMetadata(
+            filename=properties.get("filename"),
+            size_in_bytes=properties.get("size"),
+        )
+
+
+class SessionsPythonREPLTool(BaseTool):
+    """A tool for running Python code in an Azure Container Apps dynamic sessions
+    code interpreter."""
+
+    name: str = "Sessions_Python_REPL"
+    description: str = (
```

Comment:
maybe worth mentioning something about which packages are available? e.g. "You have access to all Python packages included in anaconda xyz"

Reply:
We're planning to publish a list of packages, once that's exists, maybe we can link to it from here.